modules/http4s/test/src/smithy4s/http4s/EndpointSpecificMiddlewareSpec.scala

```diff
@@ -0,0 +1,160 @@
+    )
+  }
+
+  test(
+    "server - middleware allows passing through to underlying implementation"
+  ) {
+    serverMiddlewareTest(
+      shouldFailInMiddleware = false,
+      Request[IO](Method.POST, Uri.unsafeFromString("/bob")),
+      response => {
+        response.body.compile
+          .to(Collector.supportsArray(Array))
+          .map(new String(_))
+          .map { body =>
+            expect.eql(response.status, Status.Ok) &&
+            expect.eql(body, """{"message":"Hello, bob"}""")
+          }
+      }
+    )
+  }
+
+  test("client - middleware is applied") {
+    clientMiddlewareTest(
+      shouldFailInMiddleware = true,
+      service =>
+        service.hello("bob").attempt.map { result =>
+          expect.eql(result, Left(new GenericServerError(Some("failed"))))
+        }
+    )
+  }
+
+  test("client - send request through middleware") {
+    clientMiddlewareTest(
+      shouldFailInMiddleware = false,
+      service =>
+        service.hello("bob").attempt.map { result =>
+          expect.eql(result, Right(Greeting("Hello, bob")))
+        }
+    )
+  }
+
+  private def serverMiddlewareTest(
+      shouldFailInMiddleware: Boolean,
+      request: Request[IO],
+      expect: Response[IO] => IO[Expectations]
+  )(implicit pos: SourceLocation): IO[Expectations] = {
+    val service =
+      SimpleRestJsonBuilder
+        .routes(HelloImpl)
+        .middleware(new TestMiddleware(shouldFail = shouldFailInMiddleware))
+        .make
+        .toOption
+        .get
+
+    service(request)
+      .flatMap(res => OptionT.liftF(expect(res)))
+      .getOrElse(
+        failure("unable to run request")
+      )
+  }
+
+  private def clientMiddlewareTest(
+      shouldFailInMiddleware: Boolean,
+      expect: HelloWorldService[IO] => IO[Expectations]
+  ): IO[Expectations] = {
+    val serviceNoMiddleware: HttpApp[IO] =
+      SimpleRestJsonBuilder
+        .routes(HelloImpl)
+        .make
+        .toOption
+        .get
+        .orNotFound
+
+    val client: HelloWorldService[IO] = {
+      val http4sClient = Client.fromHttpApp(serviceNoMiddleware)
+      SimpleRestJsonBuilder(HelloWorldService)
+        .client(http4sClient)
+        .middleware(new TestMiddleware(shouldFail = shouldFailInMiddleware))
+        .use
+        .toOption
+        .get
+    }
+
+    expect(client)
+  }
+
+  private object HelloImpl extends HelloWorldService[IO] {
+    def hello(name: String, town: Option[String]): IO[Greeting] = IO.pure(
+      Greeting(s"Hello, $name")
+    )
+  }
+
+  private final class TestMiddleware(shouldFail: Boolean)
+      extends EndpointSpecificMiddleware.Simple[IO] {
+    def prepareUsingHints(
+        serviceHints: Hints,
+        endpointHints: Hints
+    ): HttpApp[IO] => HttpApp[IO] = { inputApp =>
+      HttpApp[IO] { request =>
+        val hasTag: (Hints, String) => Boolean = (hints, tagName) =>
+          hints.get[smithy.api.Tags].exists(_.value.contains(tagName))
```

Comment:
first time I'm seeing anyone use this hint for anything in Scala

Reply:
Hah yeah it was just a really easy way I could think to test that the hints were being sent through properly. I don't have any real use cases for it atm. Though I guess it could come in handy for some arbitrary middleware thing 🤷